src/vm/arm64/asmhelpers.asm

```diff
@@ -1386,5 +1391,99 @@ CallHelper2
     ret lr
     LEAF_END
 
+; ------------------------------------------------------------------
+;__declspec(naked) void F_CALL_CONV JIT_Stelem_Ref(PtrArray* array, unsigned idx, Object* val)
+    LEAF_ENTRY JIT_Stelem_Ref
+    ; We retain arguments as they were passed and use x0 == array x1 == idx x2 == val
+
+    ; check for null array
+    cbz     x0, ThrowNullReferenceException
+
+    ; idx bounds check
+    ldr     x3,[x0,#ArrayBase__m_NumComponents]
+    cmp     x3, x1
+    bls     ThrowIndexOutOfRangeException
+
+    ; fast path to null assignment (doesn't need any write-barriers)
+    cbz     x2, AssigningNull
+
+    ; Verify the array-type and val-type matches before writing
+    ldr     x12, [x0] ; x12 = array MT
+    ldr     x3, [x2] ; x3 = val->GetMethodTable()
+    ldr     x12, [x12, #MethodTable__m_ElementType] ; array->GetArrayElementTypeHandle()
+    cmp     x3, x12
+    beq     JIT_Stelem_DoWrite
+
+    ; Types didnt match but allow writing into an array of objects
+    ldr     x3, =$g_pObjectClass
+    ldr     x3, [x3]  ; x3 = *g_pObjectClass
+    cmp     x3, x12   ; array type matches with Object*
+    beq     JIT_Stelem_DoWrite
+
+    ; array type and val type do not exactly match. Raise frame and do detailed match
+    b       JIT_Stelem_Ref_NotExactMatch
+
+AssigningNull
+    ; Assigning null doesn't need write barrier
+    add     x0, x0, x1, LSL #3           ; x0 = x0 + (x1 x 8) = array->m_array[idx]
+    str     x2, [x0, #PtrArray__m_Array] ; array->m_array[idx] = val
+    ret
+
+ThrowNullReferenceException
+    ; Tail call JIT_InternalThrow(NullReferenceException)
+    ldr     x0, =CORINFO_NullReferenceException_ASM
+    b       JIT_InternalThrow
+
+ThrowIndexOutOfRangeException
+    ; Tail call JIT_InternalThrow(NullReferenceException)
+    ldr     x0, =CORINFO_IndexOutOfRangeException_ASM
+    b       JIT_InternalThrow
+
+   LEAF_END 
+
+; ------------------------------------------------------------------
+; __declspec(naked) void F_CALL_CONV JIT_Stelem_Ref_NotExactMatch(PtrArray* array,
+;                                                       unsigned idx, Object* val)
+;   x12 = array->GetArrayElementTypeHandle()
+;
+    NESTED_ENTRY JIT_Stelem_Ref_NotExactMatch
+    PROLOG_SAVE_REG_PAIR           fp, lr, #-0x80!
+    SAVE_ARGUMENT_REGISTERS        sp, 0x10
```

Comment:
This will save x0-x8, but we really only need x0-x2.

Reply:
There seems to be a restriction on saving registers x0-x2 with macro PROLOG_SAVE_REG_PAIR. 
It accepts register numbers x19 to x31 only.

Reply:
@janvorli @jkotas Is this a known restriction?

Reply:
`PROLOG_SAVE_REG_PAIR` generates unwind info. Unwind info needs to be generate for non-volatile registers only (ie x19 - x31).

If you need to save scratch registers, you can same them after prolog. Or if you really need to save them during prolog, you can use `PROLOG_NOP`.